### PR TITLE
Add REST API endpoints for secrets

### PR DESCRIPTION
**Pull request by eamansour**
## Why?
See https://github.com/galasa-dev/projectmanagement/issues/1467

## Changes
- Added `GET /secrets` to get all secrets in the credentials store
- Added `GET /secrets/{secretName}` to get an individual secret
- Added `POST /secrets` to create a new secret
- Added `DELETE /secrets/{secretName}` to delete an existing secret
- Added `PUT /secrets/{secretName}` to update an existing secret, or create it if it does not exist

### CODE REVIEW

File: galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java [hunk 1]
```diff
@@ -95,8 +95,8 @@ public enum ServletErrorMessage {
...
     GAL5073_UNSUPPORTED_GALASA_SECRET_ENCODING        (5073, "E: Unsupported data encoding scheme provided. Supported encoding schemes are: [{0}]. Check that your request payload is correct and try again."),
     GAL5074_UNKNOWN_GALASA_SECRET_TYPE                (5074, "E: Unknown GalasaSecret type provided. Supported GalasaSecret types are: [{0}]. Check that your request payload is correct and try again."),
-    GAL5075_ERROR_SECRET_ALREADY_EXISTS               (5075, "E: Error occurred when trying to create a secret with the given ID. A secret with the provided ID already exists."),
-    GAL5076_ERROR_SECRET_DOES_NOT_EXIST               (5076, "E: Error occurred when trying to update a secret with the given ID. A secret with the provided ID does not exist and therefore cannot be updated."),
+    GAL5075_ERROR_SECRET_ALREADY_EXISTS               (5075, "E: Error occurred when trying to create a secret with the given name. A secret with the provided name already exists."),
+    GAL5076_ERROR_SECRET_DOES_NOT_EXIST               (5076, "E: Error occurred. A secret with the provided ID does not exist. Check that your provided secret name is correct and try again."),
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java:99]: Id and name in same message ? Aren't they both just secret names ?

File: galasa-parent/dev.galasa.framework.api.secrets/src/main/java/dev/galasa/framework/api/secrets/internal/routes/SecretDetailsRoute.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    @Override
+    public HttpServletResponse handleDeleteRequest(
+        String pathInfo,
+        QueryParameters queryParams,
+        HttpServletRequest request,
+        HttpServletResponse response
+    ) throws FrameworkException {
+        logger.info("handleDeleteRequest() entered");
+        // The provided name is implicitly validated by the route's regex pattern
+        String secretName = getSecretNameFromPath(pathInfo);
```
Comment (techcobweb) [inline galasa-parent/dev.galasa.framework.api.secrets/src/main/java/dev/galasa/framework/api/secrets/internal/routes/SecretDetailsRoute.java:124]: We get the secret name but don't validate it before we pass it onto couchdb ?

**Comment by techcobweb** [comment]
I think we said we needed a description for each credential ? I don't see that being stored away at all.

File: galasa-parent/dev.galasa.framework.api.secrets/src/main/java/dev/galasa/framework/api/secrets/internal/routes/SecretDetailsRoute.java (hunk 2 above)
Comment (eamansour) [reply galasa-parent/dev.galasa.framework.api.secrets/src/main/java/dev/galasa/framework/api/secrets/internal/routes/SecretDetailsRoute.java:124]: By the time this code is hit, the secret name has been validated by the route's regex pattern (as noted in the comment) - this currently accepts only alphanumeric characters, underscores, and dots. If the validation fails, the route handling code won't pass anything to etcd.

Happy to add more validation here, but if we'll be performing the same checks as the route's regex, the checks will always pass.

File: galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java (hunk 1 above)
Comment (eamansour) [reply galasa-parent/dev.galasa.framework.api.common/src/main/java/dev/galasa/framework/api/common/ServletErrorMessage.java:99]: Thanks for spotting this - fixed.

**Review by techcobweb — APPROVED** [review]
subject to comments being sorted out.

**Pull request merged**